harvester/node-disk-manager
# Two-phase disk unprovision

Related: harvester/harvester#1334

## Solution

Introduce Two-phase disk unprovision to prevent from data loss when user tries to remove a disk. (previously we remove disk immediately)

The new status indicates the provision phase of the device with
a Longhorn node. It's an enum type with three variants:

- `Provisioned`: Device is provisioned onto a Longhorn node as
  a disk. Longhorn replicas should be safe to schedule on it.
- `Unprovisioning`: The related disk of the device is under eviction
  from a Longhorn node. Once all replicas get evicted, the related
  disk would be removed from the node.
- `NotProvisioned`: The device is not provisioned at all. Might be
  possible to provision as a disk again.

The flow is simple:

1. User unprovisions the device/disk from the node by specifying
   `blockdevice.spec.fileSystem.provisioned = false`
2. NDM detects that a device/disk need to unprovision. To prevent
   data loss, NDM sets `evictionRequested = true`, 
   `allowScheduling = false` and disk tag `harvester-ndm-disk-remove` on the relevant disk on the longhorn
   node for rescheduling replicas to other disks. Meanwhile, the
   `blockdevice.status.provisionPhase` becomes `Unprovisioning`.
3. Once `scheduledReplica` of the disk is empty, NDM then safely
   removes the disk from the longhorn node.
   `blockdevice.status.provisionPhase` is then back to `NotProvisioned`.

## Test Plan

1. Create a 3-node cluster
2. Update blockdevice CRD and image of NDM daemonset
3. Attach some disk on a node, and then add disk onto the node in Harvester UI.
4. Create a VM and make sure your volume replicas is scheduled on the new disk (you can check it via Longhorn UI)
5. Go to node edit page and now remove the disk
6. Observe the blockdevice, you will see the phase transit from `Provisioned` to `Unprovisioning` to `NotProvisioned`. You can also check the Longhorn UI. The disk will first evict requested and not schedulable. Then it will be removed.

## TODO

- [ ] I guess the Havester UI need relevant updates?
- [ ] Update harvester/charts. Don't forget to update CRDs manifest.
